edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,631 @@
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
+                Option<string> targetAuthChain = await identitiesCache.GetAuthChain(deviceId);
+                if (!targetAuthChain.HasValue)
+                {
+                    Events.AuthorizationFail_NoAuthChain(deviceId);
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                var requestData = new DeleteModuleOnBehalfOfData($"{targetAuthChain.OrDefault()}", moduleId);
+                RegistryApiHttpResult result = await this.apiClient.DeleteModuleAsync(edgeDeviceId, requestData);
+                await this.SendResponseAsync(result.StatusCode, result.JsonContent);
+                Events.CompleteRequest(nameof(this.DeleteModuleAsync), edgeDeviceId, requestData.AuthChain, result);
+            }
+            catch (Exception ex)
+            {
+                Events.InternalServerError(nameof(this.DeleteModuleAsync), ex);
+                await this.SendResponseAsync(HttpStatusCode.InternalServerError, FormatErrorResponseMessage(ex.ToString()));
+            }
+        }
+
+        [HttpPut]
+        [Route("devices/{actorDeviceId}/modules/$edgeHub/putModuleOnBehalfOf")]
+        public async Task CreateOrUpdateModuleOnBehalfOfAsync(
+            [FromRoute] string actorDeviceId,
+            [FromHeader(Name = "if-match")] string ifMatchHeader,
+            [FromBody] CreateOrUpdateModuleOnBehalfOfData requestData)
+        {
+            try
+            {
+                Events.ReceivedOnBehalfOfRequest(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), actorDeviceId, Events.GetAdditionalInfo(requestData));
+
+                try
+                {
+                    Preconditions.CheckNonWhiteSpace(actorDeviceId, nameof(actorDeviceId));
+                    Preconditions.CheckNotNull(requestData, nameof(requestData));
+                    Preconditions.CheckNonWhiteSpace(requestData.AuthChain, nameof(requestData.AuthChain));
+                    Preconditions.CheckNotNull(requestData.Module, nameof(requestData.Module));
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(nameof(this.CreateOrUpdateModuleOnBehalfOfAsync), ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, FormatErrorResponseMessage(ex.ToString()));
+                    return;
+                }
+
+                actorDeviceId = WebUtility.UrlDecode(actorDeviceId);
+                if (!AuthChainHelpers.TryGetTargetDeviceId(requestData.AuthChain, out string targetDeviceId))
```

Comment:
>TryGetTargetDeviceId [](start = 38, length = 20)

Seems like the target identity won't ever be a module, both this method and CreateOrUpdateModuleAsync always forwards the call OnBehalfOf an Edge device.  So maybe it's better to explicit check the target ID and throw an error if it's a module.

Reply:
For registry api calls starting from iotedged, target id is always device id.  As far as I know, for on-behalf-of registry api call, authchain should not contain any module identity.  Please confirm if my understanding is correct.

Reply:
Update TryGetTargetDeviceId to return target device Id only if 1st identity is a device identity.